core/controllers/ingest.py

```diff
@@ -313,3 +299,59 @@ def save_submission_file(excel_file, submission_id):
     submission.submission_file = excel_file.stream.read()
     db.session.add(submission)
     db.session.commit()
+
+
+def get_programmes_same_round_or_older(reporting_round: int, programme_ids: list[str]) -> dict[str, str]:
+    """Returns list of programme ids of the same round or older than the reporting round.
+
+    :param reporting_round: The round currently being ingested.
+    :param programme_ids: The programme ids of the data being ingested.
+    :return: list of programme ids
+    """
+    existing_programmes_same_round_or_older = (
+        Programme.query.join(Project)
+        .join(Submission)
+        .filter(Programme.programme_id.in_(programme_ids))
+        .filter(Submission.reporting_round <= reporting_round)
+        .all()
+    )
+
+    existing_programme_ids_round_same_round_or_older = {
+        programme.programme_id: programme.id for programme in existing_programmes_same_round_or_older
+    }
+
+    return existing_programme_ids_round_same_round_or_older
+
+
+def get_programmes_newer_round(reporting_round: int, programme_ids: list[str]) -> list[str]:
+    """Returns list of programme ids of rounds newer than the reporting round.
+
+    :param reporting_round: The round currently being ingested.
+    :param programme_ids: The programme ids of the data being ingested.
+    :return: list of programme ids
+    """
+    programmes_newer_round = (
+        Programme.query.join(Project)
+        .join(Submission)
+        .filter(Programme.programme_id.in_(programme_ids))
+        .filter(Submission.reporting_round >= reporting_round)
+        .all()
+    )
+
+    programme_ids_newer_round = [programme.programme_id for programme in programmes_newer_round]
```

Comment:
Possibly, let me explore.

Reply:
with_entities returns a list of tuples, and this will require Python to unpack. It looks like queries on a single given column will always return a tuple:

https://stackoverflow.com/questions/48466959/query-for-list-of-attribute-instead-of-tuples-in-sqlalchemy

I've kept with list comprehension for now. I suppose the question worth asking is whether unpacking a tuple is less expensive than this list comprehension?